microprofile/WFLY-14987_reactive_messaging-ssl-context-kafka-connector.adoc

```diff
@@ -0,0 +1,278 @@
+
+* mailto:jstourac@redhat.com[Jan Stourac]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [x] QE
+
+=== Affected Projects or Components
+* WildFly - the Reactive Messaging integration will be enhanced to provide the functionality for this to work
+
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [ ] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+* It must be possible to use the global connector or individual method format properties to specify an Elytron provided `SSLContext` (set up via the `/subsystem=elytron/client-ssl-context` resources and its dependencies).
+** If there is no Elytron provided `SSLContext` with that name an error will be thrown on deployment
+* The examples shown above for using properties from a deployment `microprofile-config.properties`. However, the configuration is using MicroProfile Config, so other mechanisms to specify these values, such as environment variables, are also valid.
+* The indicated Elytron `SSLContext` must be present at the time of deployment, or an error will be thrown.
+* There will be a dependency from the deployment on the JBoss MSC service providing the `SSLContext`.
+
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
+* Although we encourage users to use the Elytron provided `SSLContext` mechanism, it is not an error to use the native SmallRye Reactive Messaging/Apache Kafka client properties
+* If the user has specified the `<PREFIX>wildfly.elytron.ssl.context` property in their MicroProfile Config, it is not an error to use the other Kafka native ssl properties (e.g. the ones that start with `<PREFIX>.ssl.` in the configuration). These properties will simply be ignored.
+
+== Test Plan
+At the moment, it looks like the tests for this will need to reside in QE's testsuite.
+
+The issue is that we are currently using the Spring embedded Kafka test utility for testing the Kafka integration which does not have the required hooks. testcontainers.org is also popular but requires users have a running version of Docker on their machines, which is not the case for our CI.
+
+I looked into forking the Spring utility, but noticed that the system property needed to set up JAAS (`-Djava.security.auth.login.config`) is read at JVM boot. Also, Zookeeper and Kafka need a separate config file for JAAS, and that config file must be specified on JVM startup via a system property. Hence enhancing the embedded Kafka used by the current WildFly tests does not look practical at the moment without writing a lot of process wrappers and starting them as external processes.
```

Comment:
Would it be possible to leverage [Configuration.html#setConfiguration(javax.security.auth.login.Configuration)](https://docs.oracle.com/javase/7/docs/api/javax/security/auth/login/Configuration.html#setConfiguration(javax.security.auth.login.Configuration)), as we do on [several places in WF](https://github.com/wildfly/wildfly/search?q=Configuration.setConfiguration) already?

Reply:
Now I am thinking. Why we are bothering with JAAS here, when RFE is about SSL, which does not relate to SASL authentication.

Would it be possible to write test in WF TS, when we do not care about authentication (JAAS) but just about SSL?

Reply:
JAAS is needed for the integration with RHOSAK, but perhaps that can be tested as part of one of the other RFEs now that you mention it.

The SSL context then I guess is needed to reference the truststore to accept the certificate. 

I would need to look into this Configuration thing you mention, since it is new to me. I have created https://issues.redhat.com/browse/WFLY-15283 to add the test (if possible)

Reply:
The SSL Context test PR is here https://github.com/wildfly/wildfly/pull/14665. Re JAAS, there is no special handling of that on our side, we just pass properties on to SmallRye Reactive Messaging